README
NAME

    wp-tools - tools to backup and upgrade WordPress installations

SYNOPSIS

        # simple installation from CPAN
        cpanm App::WordPressTools
    
        # view usage information
        wp-tools --help
    
        # common usage
        wp-tools upgrade|backup|restore OPTIONS...

DESCRIPTION

    WordPress Tools is a set of tools that allow you to backup, restore,
    and upgrade WordPress sites. The tools are especially useful for
    upgrading very outdated sites and for scripting the backing up and
    upgrading of many sites.

    WordPress Tools was built to make the Internet more secure. A huge
    number of websites on the Internet use WordPress
    <https://wordpress.org/>, because it's awesome, but a large portion of
    those sites do not run the latest version of WordPress which makes them
    susceptible to hacking. WordPress Tools can be used to get those sites
    up-to-date again, and that makes the whole Internet more secure.

    The code has been used in production and has already upgraded over two
    million WordPress sites, but it has little real-world testing outside
    of Linux at this point so your mileage may vary if you have a different
    operating system. Stay tuned as we add support for more platforms, and
    please contribute if you feel like it.

    The command-line program also has a lot of options for managing server
    load, so you can upgrade your sites without killing your server.

INSTALLATION

    There are several ways to install WordPress Tools to your system.

    Make sure you have WP-CLI <http://wp-cli.org/> and the other
    "DEPENDENCIES" installed.

 Using cpanm

    You can install WordPress Tools using cpanm. If you have a local perl
    (plenv, perlbrew, etc.), you can just do:

        cpanm App::WordPressTools

    to install the wp-tools executable and its dependencies. The executable
    will be installed to your perl's bin path, like
    ~/perl5/perlbrew/bin/wp-tools.

    If you're installing to your system perl, you can do:

        cpanm --sudo App::WordPressTools

    to install the wp-tools executable to a system directory, like
    /usr/local/bin/wp-tools (depending on your perl).

 Downloading just the executable

    You may also choose to download wp-tools as a single executable, like
    this:

        curl -OL https://raw.githubusercontent.com/bluehost/wp-tools/solo/wp-tools
        chmod +x wp-tools

    This executable includes all the non-core Perl module dependencies
    built-in.

 For developers

    If you're a developer and want to hack on the source, clone the
    repository and pull the dependencies:

        git clone https://github.com/bluehost/wp-tools.git
        cd wp-tools
        cpanm -n Dist::Zilla
        dzil authordeps --missing |cpanm -n
        dzil listdeps --author --develop --missing |cpanm -n

DEPENDENCIES

    WordPress Tools requires these other programs and libraries in order to
    do its work. You probably have most of these on your system already,
    and the cpanm method of installing wp-tools will handle the Perl module
    dependencies for you.

      * WP-CLI <http://wp-cli.org/>

      * GNU Coreutils <https://www.gnu.org/software/coreutils/>

      * GNU Tar <https://www.gnu.org/software/tar/>

      * awk

      * grep

      * mysql (MySQL server client)

      * procps

      * Various Perl modules (see the distributed cpanfile file for a list)

OPTIONS

    --help

      Show usage information and exit.

    --force

      If there are any resource limits in effect, ignore them and do the
      command anyway.

    --max-count=num

      Delete the oldest backups, keeping the number provided (default: 5).

    --max-dproc=num

      Refuse to execute if the maximum number of defunct processes on the
      system exceeds this number (default: 100).

    --max-load=num

      Refuse to execute if the load average exceeds this number (default:
      200).

    --max-size=num

      Refuse to back up a site if its uncompressed size on disk exceeds
      this number of bytes (default: 5368709120).

    --max-run=num

      Do not allow more than this number of concurrent system-wide
      executions of the script (default: 50).

    --min-freemem=num

      Refuse to execute if free memory has dropped below this number of
      bytes (default: 1048576).

    --wp-cli=path

      If your wp-cli program has a different name than wp or is not in your
      $PATH, you can specify the command to be used to call wp-cli.

    --username=name

      If you run wp-tools as a root user, it will drop permissions to the
      specified user and chdir to their home directory.

COMMANDS

    upgrade

      The upgrade command will upgrade your WordPress core, themes, and
      plugins to the latest version available. It takes a full backup prior
      to the update. After the update, wp-tools will do a quick check to
      try to determine that nothing broke on your site. If any failures are
      detected, wp-tools will automatically restore to the backup taken
      prior to update.

    backup

      The backup command will create a full backup, including database, of
      your WordPress installation as long as it is under the --max-size
      limit (pre-compression).

    restore

      The restore command will restore a WordPress installation from a
      backup taken using the backup command.

EXAMPLES

        # upgrade a site
        wp-tools upgrade --path=/absolute/path/to/public_html/myblog \
                         --backupdir=/absolute/path/to/myblog_backups
    
        # upgrade only plugins and themes
        wp-tools upgrade --path=/absolute/path/to/public_html/myblog  \
                         --backupdir=/absolute/path/to/myblog_backups \
                         --components=plugin,theme
    
        # backup a site
        wp-tools backup --path=/absolute/path/to/public_html/mysite \
                        --backupdir=/absolute/path/to/backups/mysite
    
        # restore a site to a previous state
        wp-tools restore --backupfile=/absolute/path/to/backups/mysite/wp_backup1428524082.tar.gz

NOTES

    Beginning with version 1.03, passing relative paths to --path,
    --backupdir, and --backupfile is deprecated. You should use absolute
    paths (i.e. that begin with a /). Relative paths are accepted but will
    print a warning to STDERR. This is because using a relative path does
    not behave as you would think (for boring historical reasons).

CONTRIBUTORS

    Many people were involved in the creation of WordPress Tools. In
    particular:

      * Matt Andersen

      * Gabriel Peery

AUTHORS

      * Seth Johnson <sj@bluehost.com>

      * Charles McGarvey <cmcgarvey@bluehost.com>

      * Garth Mortensen <gmortensen@bluehost.com>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2016-2018 by Bluehost Inc.

    This is free software, licensed under:

    The GNU General Public License, Version 2, June 1991



Changes
Revision history for App::WordPressTools (wp-tools).

{{$NEXT}}

1.03  2018-09-13 21:26:05 MDT

  * Security: Backup directories are now created 0750 (i.e. not
    world-readable) for the benefit of those with trusting umasks. If you use
    wp-tools in a shared environment, you are encouraged to upgrade.
  * Deprecated the use of relative paths as arguments to --path, --backupdir,
    and --backupfile.

1.02  2016-12-12 16:00:29 MST

  * Reduce chance of leaving site directory in an inconsistent state if the
    script is killed at the wrong time.
  * Fix broken logic for applying backup skips.

1.01  2016-02-19 15:51:50 MST

  * Fixed a bug with complete version number not being preserved.
  * Fixed a bug that made --force ineffective against the disk space check.
  * Fixed a bug related to innacurate admin page checking on upgrade.
  * Fixed test scripts to invoke the correct perl.
  * Added documentation for how to contribute.
  * Added Travis CI configuration for automated testing (thanks Jason Hall).
  * Fixed various issues with the dzil config and maintenance scripts.
  * Replaced the File::Slurp dependency with File::Slurper.
  * Replaced the LWP::UserAgent dependency with HTTP::Tiny.
  * Replaced the Digest::SHA dependency with Digest::MD5.
  * Made the solo distribution be fatpacked for easier installation by curl.

1.0  2016-02-11 14:41:53 MST

  * First release of wp-tools to the world.



maint/fatpack.pl
#!/usr/bin/env perl

# This script creates a fatpacked version of wp-tools. Much of this code was
# inspired by or blatantly copied from cpanminus build scripts, written by
# Tatsuhiko Miyagawa.

use warnings FATAL => 'all';
use strict;
use autodie ':all';

use App::FatPacker ();
use Cwd;
use File::Find;
use File::Path;
use File::pushd;
use Module::CoreList;


my $script_name = 'script/wp-tools';


BEGIN {
    # IO::Socket::IP requires newer Socket, which is C-based
    $ENV{PERL_HTTP_TINY_IPV4_ONLY} = 1;
}

END {
    no autodie;
    unlink('wp-tools.tmp');
    rmtree('.fatpack-build');
    rmtree('fatlib');
}


sub find_requires {
    my $file = shift;

    my %requires;
    open my $in, "<", $file;
    while (<$in>) {
        /^\s*(?:use|require) (\S+)[^;]*;\s*$/
          and $requires{$1} = 1;
    }

    keys %requires;
}

sub mod_to_pm {
    local $_ = shift;
    s!::!/!g;
    "$_.pm";
}

sub pm_to_mod {
    local $_ = shift;
    s!/!::!g;
    s/\.pm$//;
    $_;
}

sub in_lib {
    my $file = shift;
    -e "lib/$file";
}

sub is_core {
    my $module = shift;
    exists $Module::CoreList::version{5.008001}{$module};
}

sub exclude_modules {
    my($modules, $except) = @_;
    my %exclude = map { $_ => 1 } @$except;
    [ grep !$exclude{$_}, @$modules ];
}

sub pack_modules {
    my($path, $modules, $no_trace) = @_;

    $modules = exclude_modules($modules, $no_trace);

    my $packer = App::FatPacker->new;
    my @requires = grep !is_core(pm_to_mod($_)), grep /\.pm$/, split /\n/,
      $packer->trace(use => $modules, args => ['-e', 1]);
    push @requires, map mod_to_pm($_), @$no_trace;

    my @packlists = $packer->packlists_containing(\@requires);
    for my $packlist (@packlists) {
        print "Packing $packlist\n";
    }
    $packer->packlists_to_tree($path, \@packlists);
}

sub make_fatlib {
    my @modules = grep !in_lib(mod_to_pm($_)), find_requires($script_name);

    pack_modules(cwd . '/fatlib', \@modules, []);

    use Config;
    print "Remove fatlib/$Config{archname}\n";
    rmtree("fatlib/$Config{archname}");
    rmtree("fatlib/POD2");

    my $want = sub {
        if (/\.pod$/) {
            print "Remove $_\n";
            unlink $_;
        }
    };

    find({ wanted => $want, no_chdir => 1 }, 'fatlib');
}


sub generate_file {
    my($base, $target, $fatpack) = @_;

    open my $in,  "<", $base;
    open my $out, ">", "$target.tmp";

    print STDERR "Generating $target from $base\n";

    while (<$in>) {
        s|^#!perl|#!/usr/bin/env perl|;
        s|^# FATPACK.*|$fatpack|;
        print $out $_;
    }

    close $out;

    eval { unlink $target };
    rename "$target.tmp", $target;
}

sub make_script {
    mkdir '.fatpack-build';
    system qw(cp -r fatlib lib .fatpack-build/);

    my $fatpack_compact = do {
        my $dir = pushd '.fatpack-build';

        my @files;
        my $want = sub {
            push @files, $_ if /\.pm$/;
            if (/\.pod$/) {
                print "Remove $_\n";
                unlink $_;
            }
        };

        find({ wanted => $want, no_chdir => 1 }, 'fatlib', 'lib');
        system 'perlstrip', '--cache', '-v', @files;

        `fatpack file`;
    };

    my $filename = $script_name;
    $filename =~ s!^.*/!!;

    generate_file($script_name, $filename, $fatpack_compact);
    chmod 0755, $filename;
}


make_fatlib;
make_script;



maint/branch_dist.pl
#!/usr/bin/env perl

# This script creates a separate tree and branch for just the dist.

use strict;
use warnings;
use autodie;

use File::Copy;
use File::Path qw(make_path remove_tree);
use String::ShellQuote;


my $version = shift or die 'Need version';
my $distdir = shift or die 'Need distdir';

my $branch_name = 'dist';

my $branch_oldref = '';
my $branch_oldref_origin = '';

open(my $fh, '-|', qw{git show-ref}, $branch_name);
while (my $line = <$fh>) {
    chomp $line;
    my ($hash, $ref) = split(/\s+/, $line);
    $branch_oldref = $hash          if $ref eq "refs/heads/$branch_name";
    $branch_oldref_origin = $hash   if $ref eq "refs/remotes/origin/$branch_name";
}
if ($branch_oldref_origin && $branch_oldref ne $branch_oldref_origin) {
    # reset local branch
    system(qw{git branch -f}, $branch_name, "origin/$branch_name");
    $branch_oldref = $branch_oldref_origin
}

open(my $in,  '<', '.travis.yml');
open(my $out, '>', "$distdir/.travis.yml");
while (my $line = <$in>) {
    $line =~ s/.*(?:cpanm|dzil).*//s;
    $line =~ s/^install:/install:\n   - "cpanm --quiet --installdeps --notest ."/;
    $line =~ s/^script:/script:\n   - "perl Makefile.PL && make test"/;
    print $out $line;
}
close($in);
close($out);

my $commit_msg = shell_quote("Release $version");

system(qw{git add -f}, $distdir);
my $tree_ref = `git write-tree --prefix=$distdir/`;
chomp $tree_ref;

system(qw{git reset}, $distdir);

my $branch_oldref_safe = shell_quote($branch_oldref);
my $tree_ref_safe = shell_quote($tree_ref);
my $parent = $branch_oldref ? "-p $branch_oldref_safe" : '';
my $commit_ref = `git commit-tree -m $commit_msg $parent $tree_ref_safe`;
chomp $commit_ref;

system(qw(git branch -f), $branch_name, $commit_ref);
system(qw(git tag -a -m), "Version $version", "$branch_name-$version", $commit_ref);



maint/branch_solo.pl
#!/usr/bin/env perl

# This script prepares the wp-tools script for standalone use and puts it in a new branch.

use strict;
use warnings;
use autodie;

use File::Copy;
use File::Path qw(make_path remove_tree);
use String::ShellQuote;


my $version = shift or die 'Need version';
my $distdir = shift or die 'Need distdir';

my $branch_name = 'solo';

my $branch_oldref = '';
my $branch_oldref_origin = '';

open(my $fh, '-|', qw{git show-ref}, $branch_name);
while (my $line = <$fh>) {
    chomp $line;
    my ($hash, $ref) = split(/\s+/, $line);
    $branch_oldref = $hash          if $ref eq "refs/heads/$branch_name";
    $branch_oldref_origin = $hash   if $ref eq "refs/remotes/origin/$branch_name";
}
if ($branch_oldref_origin && $branch_oldref ne $branch_oldref_origin) {
    # reset local branch
    system(qw{git branch -f}, $branch_name, "origin/$branch_name");
    $branch_oldref = $branch_oldref_origin
}

my $commit_msg = shell_quote("Release $version");

my $solodir = "solo_branch.$$";
make_path($solodir);

use Config;
system($Config{'perlpath'}, qw{maint/fatpack.pl});
move('wp-tools', "$solodir/wp-tools");

copy("$distdir/README", "$solodir/README");
copy("$distdir/CONTRIBUTING", "$solodir/CONTRIBUTING");

system(qw{git update-index --add}, glob("$solodir/*"));
my $tree_ref = `git write-tree --prefix=$solodir/`;
chomp $tree_ref;

system(qw{git reset});
remove_tree($solodir);

my $branch_oldref_safe = shell_quote($branch_oldref);
my $tree_ref_safe = shell_quote($tree_ref);
my $parent = $branch_oldref ? "-p $branch_oldref_safe" : '';
my $commit_ref = `git commit-tree -m $commit_msg $parent $tree_ref_safe`;
chomp $commit_ref;

system(qw(git branch -f), $branch_name, $commit_ref);
system(qw(git tag -a -m), "Version $version", "$branch_name-$version", $commit_ref);



Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.012.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "tools to backup and upgrade WordPress installations",
  "AUTHOR" => "Seth Johnson <sj\@bluehost.com>, Charles McGarvey <cmcgarvey\@bluehost.com>, Garth Mortensen <gmortensen\@bluehost.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "App-WordPressTools",
  "EXE_FILES" => [
    "script/wp-tools"
  ],
  "LICENSE" => "gpl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "App::WordPressTools",
  "PREREQ_PM" => {
    "Digest::MD5" => 0,
    "Fcntl" => 0,
    "File::Find" => 0,
    "File::Path" => 0,
    "File::Slurper" => 0,
    "Getopt::Long" => 0,
    "HTTP::Tiny" => 0,
    "String::ShellQuote" => 0,
    "sigtrap" => 0
  },
  "TEST_REQUIRES" => {
    "Capture::Tiny" => 0,
    "Test::More" => 0
  },
  "VERSION" => "1.03",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => 0,
  "Digest::MD5" => 0,
  "Fcntl" => 0,
  "File::Find" => 0,
  "File::Path" => 0,
  "File::Slurper" => 0,
  "Getopt::Long" => 0,
  "HTTP::Tiny" => 0,
  "String::ShellQuote" => 0,
  "Test::More" => 0,
  "sigtrap" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
